docs/feature_rawhid.md

```diff
@@ -0,0 +1,119 @@
+# Raw HID
+
+Raw HID allows for bidirectional communication between keyboard and a host computer over HID interface. It allows for 32 byte bidirectional packet transfer in QMK.
+
+There are two main components to getting raw HID working with your keyboard.
+
+## Keyboard firmware
+
+The implementation is fairly straightforward for the firmware.
+In your `rules.mk` add:
+
+```make
+RAW_ENABLE = yes
+```
+
+In your `keymap.c` include `"raw_hid.h"` and implement the following:
+
+```C
+void raw_hid_receive(uint8_t *data, uint8_t length) {
+    // Your code goes here. data is the packet received from host.
+}
+```
+
+The `"raw_hid.h"` header also includes `void raw_hid_send(uint8_t *data, uint8_t length);`
+which allows sending packets from keyboard to host. It can be used for debugging when building your host by returning all data back to host to test communication.
+
+```C
+void raw_hid_receive(uint8_t *data, uint8_t length) {
+    raw_hid_send(data, length);
+}
+```
+
+All Raw HID packets by default have size `RAW_EPSIZE` which is `32 bytes`, the actual length is determined by `uint8_t length` both when sending and receiving.
+
+Make sure to flash raw enabled firmware before proceeding with working on the host side.
+
+## Host (PC/Mac/Linux)
+
+This is the more complicated part as it will require some digging.
+
+To connect your host computer to your keyboard with raw HID you need four pieces of information about your keyboard:
+
+1. Vendor ID
+2. Product ID
+3. Usage Page
+4. Usage
+
+The first two can easily be found in your keyboard's `config.h` in the keyboard's main directory under `VENDOR_ID` and `PRODUCT_ID`.
+
+The second two are slightly harder and can vary by vendor. Most of the time though for most QMK keyboards **usage page is `0xFF60`** and **usage is `0x0061`**.
+
+#### The harder way
+
+If these values don't work for you, you can easily discover the usage page and usage for your keyboard by finding the values for `tag 4` for usage page and `tag 8` for usage in the hid interface descriptor for your keyboard by using any form of hid dump or any hid library to find the descriptor and extract the values from the tags.
+ 
+The easy way to do this for linux and mac users is to use teensy raw HID [host implementation](https://www.pjrc.com/teensy/rawhid.html "PJRC Teensy") to find the two tags' values. It is also recommended to use this host implementation as a starting point for your own host on any platform.
```

Comment:
I would not recommend using the Teensy code anymore as it doesn't build OOB on macOS (the make config is outdated) and generally seems rather bitrotted now.

Reply:
yeah, it is quite old. We should present some skeleton host for testing purposes though. There is [this](https://github.com/rhruiz/hid_send) in C, and [this](https://github.com/BlankSourceCode/qmk-hid-display) with node-hid, this one is somewhat similar to the way VIA connects. Or, we could use the VIA code, refer to the specific file in the repo, but it's so not a skeleton.

Reply:
I'd imagine documenting the APIs for the moment will suffice -- we can always open up a future PR for a skeleton app.

Reply:
Agreed. Let's get API documentation in and we can do implementation and client stuff later.